src/comm/TcpAcceptor.cc

```diff
@@ -271,33 +272,41 @@ Comm::TcpAcceptor::acceptOne()
 
     /* Accept a new connection */
     ConnectionPointer newConnDetails = new Connection();
-    const Comm::Flag flag = oldAccept(newConnDetails);
-
-    if (flag == Comm::COMM_ERROR) {
-        // A non-recoverable error; notify the caller */
-        debugs(5, 5, "non-recoverable error:" << status() << " handler Subscription: " << theCallSub);
-        if (intendedForUserConnections())
-            logAcceptError(newConnDetails);
-        notify(flag, newConnDetails);
-        // XXX: not under async job call protections
-        mustStop("Listener socket closed");
+    try {
+        if (acceptInto(newConnDetails)) {
+            Assure(newConnDetails->isOpen());
+            CallBack(newConnDetails, [&] {
+                debugs(5, 5, "Listener: " << conn <<
+                       " accepted new connection " << newConnDetails <<
+                       " handler Subscription: " << theCallSub);
+                notify(Comm::OK, newConnDetails);
+            });
+        } else {
+            debugs(5, 5, "try later: " << conn << " handler Subscription: " << theCallSub);
+            newConnDetails->close(); // paranoid manual closure (and may already be closed)
```

Comment:
This explicit closure is not really needed because we are not sharing newConnDetails with anybody in this case. I will remove it if others think it should be removed instead:

```suggestion
```

Reply:
Please do close(). "sharing" is irrelevant. The need to close is about leaving the TCP stack with a hung socket. A Comm::Connection always needs to be closed at the point where it becomes known to be unusable.

Reply:
> Please do keep the close().

Sure, that is the current plan.

> The need to close is about leaving the TCP stack with a hung socket.

Whether we remove this explicit close() or not, this connection will be closed before we leave this function (because this connection object is not shared). However, I think keeping this explicit close() is slightly better than removing it (because we cannot be sure whether this connection object would never be shared with some other job in some future code, even though such sharing would be a bug -- an explicit close() with a comment clarifies our thinking/assumptions a bit).

> A Comm::Connection always needs to be closed at the point where it becomes known to be unusable.

I disagree with the "always" part. In many cases, the connection should indeed be explicitly closed at that point. In some cases, it should be explicitly closed, but a bit later (e.g., when we reach some common cleanup code in the same function). In some other cases, it is better to rely on RAII and other automated mechanisms than explicit closures. Fortunately, we do not need to agree on this for this PR to be merged.